src/convert.rs

```diff
@@ -1,69 +1,56 @@
 //! Conversion between bls12_381 types and bytes.
 
-use crate::error::{Error, Result};
-use group::{CurveAffine, CurveProjective, EncodedPoint};
+use blst::{blst_fr, blst_fr_from_scalar, blst_scalar};
 
-use crate::{PK_SIZE, SIG_SIZE, SK_SIZE};
-use ff::PrimeField;
-use pairing::bls12_381::{Fr, FrRepr, G1Affine, G2Affine, G1, G2};
+use crate::{
+    error::{Error, Result},
+    Fr, DST, G1, G2, PK_SIZE, SIG_SIZE, SK_SIZE,
+};
 
-/// Convert big endian bytes to bls12_381::Fr
-pub fn fr_from_be_bytes(bytes: [u8; SK_SIZE]) -> Result<Fr> {
-    let mut le_bytes = bytes;
-    le_bytes.reverse();
-    let mut fr_u64s = [0u64; 4];
-    for i in 0..4 {
-        let mut next_u64_bytes = [0u8; 8];
-        for j in 0..8 {
-            next_u64_bytes[j] = le_bytes[i * 8 + j];
-        }
-        fr_u64s[i] = u64::from_le_bytes(next_u64_bytes);
-    }
-    Ok(Fr::from_repr(FrRepr(fr_u64s))?)
-}
-
-/// Convert bls12_381::Fr to big endian bytes
-pub fn fr_to_be_bytes(fr: Fr) -> [u8; SK_SIZE] {
-    let mut bytes = [0u8; SK_SIZE];
-    // iterating 4 u64s which are in order suiting little endian bytes
-    // and must be reversed to get big endian bytes
-    fr.into_repr().0.iter().enumerate().for_each(|(le_i, n)| {
-        let be_i = 4 - le_i - 1;
-        n.to_be_bytes().iter().enumerate().for_each(|(j, byte)| {
-            bytes[8 * be_i + j] = *byte;
-        });
-    });
-    bytes
+pub(crate) fn derivation_index_into_fr(index: &[u8]) -> Fr {
+    hash_to_field(index)
 }
 
-/// Convert big endian bytes to bls12_381::G1
-pub fn g1_from_be_bytes(bytes: [u8; PK_SIZE]) -> Result<G1> {
-    let mut compressed: <G1Affine as CurveAffine>::Compressed = EncodedPoint::empty();
-    compressed.as_mut().copy_from_slice(&bytes);
-    let opt_affine = compressed.into_affine().ok();
-    let projective = opt_affine.ok_or(Error::InvalidBytes)?.into_projective();
-    Ok(projective)
+/// Generates a scalar as described in IETF hash to curve
+/// https://www.ietf.org/archive/id/draft-irtf-cfrg-hash-to-curve-10.html#name-hashing-to-a-finite-field-2
+/// There are two main candidates for converting arbitrary bytes to scalar:
+/// * hash to field as described in hash to curve
+/// * bls signature keygen based on hkdf
+/// BLS Signature spec has strong recommendations about IKM which don't apply when deriving indexes.
+/// Hash to field is also slightly faster than HKDF.
+fn hash_to_field<T: AsRef<[u8]>>(msg: T) -> Fr {
+    // TODO IC remove unwrap here
+    let scalar = blst_scalar::hash_to(msg.as_ref(), DST).unwrap();
```

Comment:
maybe this fn should return a Result<> if the unwrap cannot be removed as-is.    If too large a change, it could be done in a follow-up PR.

Reply:
There doesn't seem to be any practical way this can fail to unwrap.

Some implementation details and background info:

`blst_scalar::hash_to` is defined in [lib.rs#L161](https://github.com/supranational/blst/blob/03d70d92e9b206fb9b6354bd7040ec0eb946535c/bindings/rust/src/lib.rs#L161). It first hashes the input bytes on L166 into a 48 byte array, then calls `blst_scalar_from_be_bytes` on L174 (which will `mod` those bytes with `Fr::MAX`).

`blst_scalar_from_be_bytes` is defined in [exports.c#L511](https://github.com/supranational/blst/blob/50bf48131dcffdbc76ee456a988772b664851f37/src/exports.c#L511) and fails *only* if the output is `Fr::zero` on L535. This can happen if the input is `[0u8; 32]` or any multiple of `Fr::MAX+1`. Our input is the result of a hash so it's very unlikely to give a result of exactly zero.

In reality the unwrap will never fail.

All the same, we can and probably should wrap it in a Result. Returning zero (or panicking) would be very bad since we'd get a black hole for child keys: `parent_key * zero -> zero`.

I'll add this change to this PR.

Reply:
Hash to field now returns a Result. It will only return an error in the extremely unlikely event the input hashes to Fr::zero.

The only place hash to field is used is in `derivation_index_into_fr`, which *doesn't* return an error. Instead it repeats the hash to field a second time (using Fr::zero bytes as the input) to give a non-zero value. This seems like a valid approach to me because sending the consumer an error isn't helpful, what can they really do about that error? Better to be consistent and avoid each consumer having their own different way of handling it. It seems preferable to me that `derive_child` always returns a valid child key, rather than sometimes not.